src/web/services/admin/semesters/view_enrollments.rs

```diff
@@ -0,0 +1,185 @@
+        aweb::get().to(enrollments_page_index),
+    );
+}
+
+/// Try to get the pagination bar to use based on the api data.
+/// Panics if `current_page` is 0.
+fn get_page_numbers(api_response: &Value, current_page: u64) -> Option<PaginationInfo> {
+    api_response
+        // Check for the JSON field user_count
+        .get("user_count")?
+        // With field aggregate
+        .get("aggregate")?
+        // With field count
+        .get("count")?
+        // As an unsigned integer
+        .as_u64()
+        // Convert to pagination info
+        .and_then(|count| PaginationInfo::new(count, PER_PAGE as u64, current_page))
+}
+
+// download page for enrollments csv file.
+// When user access to this page, a csv file will be created and written at /tmp/.
+#[get("/download/enrollments/{semester_id}")]
+pub async fn export_to_csv(
+    Path(semester_id): Path<String>,
+) -> Result<HttpResponse, TelescopeError> {
+    let query_response = EnrollmentsLookup::get(semester_id.clone()).await?;
+    let mut buffer = Vec::new();
+    // scope to ensure writer is dropped after its done, so we can use the buffer
+    {
+        let mut wtr = WriterBuilder::new().from_writer(&mut buffer);
+        let api_data = query_response.enrollments;
+        wtr.serialize(api_data).map_err(|e| {
+            TelescopeError::ise(format!(
+                "There was an issue writing the data to CSV: {:?}",
+                e
+            ))
+        })?;
+        wtr.flush().map_err(|e| {
+            TelescopeError::ise(format!(
+                "There was an issue finalizing the CSV file: {:?}",
+                e
+            ))
+        })?;
+    }
+    let resp = HttpResponse::Ok()
+        .set_header(header::CONTENT_TYPE, "text/csv")
+        .set_header(
+            header::CONTENT_DISPOSITION,
+            ContentDisposition {
+                disposition: DispositionType::Attachment,
+                parameters: vec![DispositionParam::Filename(format!(
+                    "enrollments-{}.csv",
+                    semester_id
+                ))],
+            },
+        )
+        .body(buffer);
+    Ok(resp)
+}
+
+pub async fn enrollments_page_index(
+    req: HttpRequest,
+    identity: Identity,
+    Path((semester_id, page)): Path<(String, u32)>,
+    Query(query): Query<EnrollmentPageQuery>,
+) -> Result<Page, TelescopeError> {
+    // Resolve the page number from the request
+    let mut page_num = page;
+    if page_num >= 1 {
+        page_num -= 1;
+    } else {
+        page_num = 0;
+    }
+
+    // Get the API data by sending one of the enrollment page queries.
+    let semester = Semester::get_by_id(semester_id.clone()).await?;
+    let query_response =
+        UserEnrollmentLookup::get_by_id(page_num, query.search.clone(), semester_id.clone())
+            .await?;
+    let enrollments = query_response.enrollments.clone();
+    let enrollment_data = serde_json::to_value(enrollments).unwrap();
```

Comment:
Do you mind changing all these `unwrap()`s to use the map_err method I did above in my changes?